### PR TITLE
azure config store for Spring Property Source

### PR DESCRIPTION
**Pull request by sophiaso**
## Description

- Azure Config Store as remote Spring Property Source
- Sample code to use the Azure Config store

### CODE REVIEW

File: spring-cloud-azure-samples/spring-cloud-azure-config-sample/pom.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        <version>1.0.0.BUILD-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>spring-cloud-azure-config-sample</artifactId>
+    <name>Spring Cloud Azure Config Sample</name>
+    <dependencies>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
+            <artifactId>spring-cloud-azure-config</artifactId>
```
Comment (xscript) [inline spring-cloud-azure-samples/spring-cloud-azure-config-sample/pom.xml]: We will have a `spring-cloud-starter-azure-config` package, right?

File: spring-cloud-azure-config/pom.xml [hunk 2]
```diff
@@ -45,5 +57,9 @@
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-databind</artifactId>
         </dependency>
+        <dependency>
```
Comment (warrenzhu25) [inline spring-cloud-azure-config/pom.xml]: This should already include `spring-boot-autoconfigure`

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import com.microsoft.azure.spring.cloud.config.domain.KeyValueItem;
+import org.springframework.core.env.EnumerablePropertySource;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class AzureConfigPropertySource extends EnumerablePropertySource<ConfigServiceOperations> {
+    private String context;
```
Comment (warrenzhu25) [inline spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: `final` ?

File: spring-cloud-azure-config/pom.xml (hunk 2 above)
Comment (sophiaso) [reply spring-cloud-azure-config/pom.xml]: As it's `optional` in `spring-boot-autoconfigure`, so I think we have to explicitly include it here, though in this repo's case, these jackson library has also been transitively dependent by other azure modules.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceLocator.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        List<String> profiles = Arrays.asList(env.getActiveProfiles());
+
+        /* Generate which contexts(key prefixes) will be used for key-value items search
+           If key prefix is empty, default context is: application, current application name is: foo,
+           active profile is: dev, separator is: /
+           Will generate these contexts: /application, /application/dev, /foo, /foo/dev
+         */
+        generateContexts(this.properties.getDefaultContext(), profiles);
```
[the remark this answers is not in the document]
Comment (sophiaso) [reply spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceLocator.java]: As it's only private accessed, so should have much problem, but will update.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceLocator.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        if (!StringUtils.hasText(applicationName)) {
+            applicationName = env.getProperty(SPRING_APP_NAME_PROP);
+        }
+
+        List<String> profiles = Arrays.asList(env.getActiveProfiles());
+
+        /* Generate which contexts(key prefixes) will be used for key-value items search
+           If key prefix is empty, default context is: application, current application name is: foo,
+           active profile is: dev, separator is: /
+           Will generate these contexts: /application, /application/dev, /foo, /foo/dev
```
[the remark this answers is not in the document]
Comment (sophiaso) [reply spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceLocator.java]: Good question, need more discussion on setting rules for this case.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    public String getCredential() {
+        return splitConnectionStr()[1].replaceFirst("Id=", "");
+    }
+
+    public String getSecret() {
+        return splitConnectionStr()[2].replaceFirst("Secret=", "");
+    }
+
+    private String[] splitConnectionStr() {
```
[the remark this answers is not in the document]
Comment (sophiaso) [reply spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java]: Updated to initialize only once.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java (hunk 3 above)
Comment (sophiaso) [reply spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: Updated

File: spring-cloud-azure-config/pom.xml (hunk 2 above)
Comment (sophiaso) [reply spring-cloud-azure-config/pom.xml]: Removed extra dependency.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public String getSecret() {
+        return secret;
+    }
+
+    @PostConstruct
+    private void validateAndInit() {
+        String[] items = connectionString.split(";");
+        Assert.isTrue(items.length == 3, "Connection string should contain three items split by ;.");
```
Comment (xscript) [inline spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java]: We shouldn't care about how many items are in the connection string. All we care is the connection string contains `endpoint`, `id` and `secret`, right? Extra key vault pairs can be safely ignored. And our package will still work if service team decides to add more fields in the connection string.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+        this.endpoint = items[0].replaceFirst("Endpoint=", "");
```
Comment (xscript) [inline spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java]: We can not presume the order of the properties.

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    @PostConstruct
+    private void validateAndInit() {
+        String[] items = connectionString.split(";");
+        for (String item : items) {
+            if (!StringUtils.hasText(item)) {
+                continue;
+            }
+
+            if (item.startsWith(ENDPOINT_PREFIX)) {
```
Comment (xscript) [inline spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java]: Is `startsWith` case sensitive?

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    private String label;
+
+    // Profile separator for the key name, e.g., /foo-app_dev/db.connection.key
+    @NotEmpty
+    @Pattern(regexp = "^[a-zA-Z0-9_@]+$")
+    private String profileSeparator = "_";
+
+    // Values extracted from connection string
+    private String endpoint;
+    private String credential;
```
Comment (xscript) [inline spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java]: Why don't we name it `id` ? Align with `.NET`?

File: spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            if (item.startsWith(ENDPOINT_PREFIX)) {
+                this.endpoint = item.substring(ENDPOINT_PREFIX.length());
+            } else if (item.startsWith(ID_PREFIX)) {
+                this.credential = item.substring(ID_PREFIX.length());
+            } else if (item.startsWith(SECRET_PREFIX)) {
+                this.secret = item.substring(SECRET_PREFIX.length());
+            }
+        }
+
+        Assert.hasText(this.endpoint, "Endpoint should not be null or empty.");
```
Comment (xscript) [inline spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigProperties.java]: The error message is not clear enough. How about `Endpoint property should not be null or empty in the connection string of Azure Config Service.` ?

**Pull request merged**